src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.Stream.cs

```diff
@@ -0,0 +1,347 @@
+            private byte[] buffer;
+            int numBytesNotWrittenFromPreviousChunk = 0;
+
+            /// <summary>
+            /// Initializes a new instance of the ODataUtf8JsonWriterStream class with the specified ODataUtf8JsonWriter.
+            /// </summary>
+            /// <param name="writer">The OData UTF-8 JSON writer to write to.</param>
+            internal ODataUtf8JsonWriteStream(ODataUtf8JsonWriter writer)
+            {
+                this.jsonWriter = writer;
+            }
+
+            public override bool CanRead => false;
+
+            public override bool CanSeek => false;
+
+            public override bool CanWrite => true;
+
+            /// <summary>
+            /// Gets the length in bytes of the stream. This is not supported by this stream.
+            public override long Length => throw new NotSupportedException();
+
+            /// <summary>
+            /// Gets or sets the position within the stream. This is not supported by this stream.
+            /// </summary>
+            public override long Position { get => throw new NotSupportedException(); set => throw new NotSupportedException(); }
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream synchronously.
+            /// </summary>
+            public override void Flush()
+            {
+                this.jsonWriter.Flush();
+            }
+
+            /// <summary>
+            /// Flushes any buffered data to the underlying stream asynchronously.
+            /// </summary>
+            /// <param name="cancellationToken">A A cancellation token to observe while waiting for the flush operation to complete.</param>
+            /// <returns>A task representing the asynchronous flush operation.</returns>
+            public override async Task FlushAsync(CancellationToken cancellationToken)
+            {
+                await this.jsonWriter.FlushAsync().ConfigureAwait(false);
+            }
+
+            /// <summary>
+            /// Disposes the object.
+            /// </summary>
+            /// <param name="disposing">True if called from Dispose; false if called form the finalizer.</param>
+            protected override void Dispose(bool disposing)
+            {
+                if (this.numBytesNotWrittenFromPreviousChunk > 0)
+                {
+                    // If there are unprocessed bytes, encode and write them as the final block.
+                    ReadOnlySpan<byte> bytesNotProcessedFromPreviousChunk = this.buffer.AsSpan().Slice(0, this.numBytesNotWrittenFromPreviousChunk);
+
+                    this.jsonWriter.Base64EncodeAndWriteChunk(bytesNotProcessedFromPreviousChunk, isFinalBlock: true, out numBytesNotWrittenFromPreviousChunk);
+                    this.numBytesNotWrittenFromPreviousChunk = 0;
+                }
+
+                if (this.buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(this.buffer);

```

Comment:
Should we set `this.buffer` to null after returning it?

Reply:
I don't think it's necessary at this point or there will ever be a point where the buffer is accessed while null.